### PR TITLE
TRUNK-5875:Unit test files left in place

### PR DESCRIPTION
**Pull request by jwnasambu**
## Description of what I changed
I worked on the BaseContextSensitiveTest for JUnit 4 and JUnit5 to ensure it properly clean-up a directory after running the unit tests.

## Issue I worked on
https://issues.openmrs.org/browse/TRUNK-5875

### CODE REVIEW

File: api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java [hunk 1]
```diff
@@ -64,7 +66,8 @@
 	
 	private DatabaseUpgradeTestUtil upgradeTestUtil;
 	
-	private static File testAppDataDir;
+	@ClassRule
+	public static TemporaryFolder testAppDataDir = TemporaryFolder.builder().assureDeletion().build();
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java:70]: you need to remove 125-128. please remember that JUnit is now creating and deleting the directory. So we do not need any code that creates/deletes athw testAppDataDir anymore

File: api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java [hunk 2]
```diff
@@ -345,19 +349,9 @@ public Properties getRuntimeProperties() {
 		// we don't want to try to load core modules in tests
 		runtimeProperties.setProperty(ModuleConstants.IGNORE_CORE_MODULES_PROPERTY, "true");
 		
-		try {
```
Comment (tendomart) [inline api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:348]: @jwnasambu  you are trying to use a variable of a non existing or deleted object.

File tempappdir = File.createTempFile("appdir-for-unit-tests-", ""); was deleted.

Comment (tendomart) [reply api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:348]: I believe that's why Travis is throwing   "[ERROR] /home/travis/build/openmrs/openmrs-core/api/src/test/java/org/openmrs/test/BaseContextSensitiveTest.java:[361,33] cannot find symbol

  symbol:   method getAbsolutePath() "

Comment (jwnasambu) [reply api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:348]: @tendomart thanks for the review but I stand to be corrected but I think we need .getRoot()
        .getAbsolutePath()); to sort out the error

File: api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java [hunk 3]
```diff
@@ -131,6 +132,9 @@
 	 */
 	protected static Properties runtimeProperties;
 	
+	@TempDir
+	public static File tempappdir;
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:136]: @tendomart it dies exist its defined here

File: api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java (hunk 2 above)
Comment (teleivo) [reply api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:348]: you can also read https://junit.org/junit5/docs/current/user-guide/#writing-tests-built-in-extensions-TempDirectory ; it also supports the newer java.nio.file.Path on which you can call https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#toAbsolutePath()

I suggest you try using the type java.nio.file.Path.

Comment (jwnasambu) [reply api/src/test/java/org/openmrs/test/jupiter/BaseContextSensitiveTest.java:348]: @teleivo thanks let me do it.

File: api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java [hunk 4]
```diff
@@ -119,12 +122,8 @@ public static void createOrderEntryUpgradeFileWithTestData(String propString) th
...
 	@BeforeAll
 	public static void beforeClass() throws IOException {
-		testAppDataDir = File.createTempFile("appdir-for-unit-tests", "");
-		testAppDataDir.delete();// so we can make turn it into a directory
-		testAppDataDir.mkdir();
-		
-		System.setProperty(OpenmrsConstants.APPLICATION_DATA_DIRECTORY_RUNTIME_PROPERTY, testAppDataDir.getAbsolutePath());
-		OpenmrsUtil.setApplicationDataDirectory(testAppDataDir.getAbsolutePath());
+		System.setProperty(OpenmrsConstants.APPLICATION_DATA_DIRECTORY_RUNTIME_PROPERTY, testAppDataDir.getRoot().getAbsolutePath());
+		OpenmrsUtil.setApplicationDataDirectory(testAppDataDir.getRoot());
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java:126]:
```suggestion
		OpenmrsUtil.setApplicationDataDirectory(testAppDataDir.getRoot().getAbsolutePath());
```

Comment (jwnasambu) [reply api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java:126]: @ibacher thanks so much! let me try out

File: api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java [hunk 5]
```diff
@@ -64,7 +66,8 @@
 	
 	private DatabaseUpgradeTestUtil upgradeTestUtil;
 	
-	private static File testAppDataDir;
+	@ClassRule
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java:69]: this test class is a JUnit 5 one so it needs the JUnit 5 `@TempDir`
please try the java.nio.file.Path here as well

File: api/src/test/java/org/openmrs/test/BaseContextSensitiveTest.java [hunk 6]
```diff
@@ -352,19 +357,9 @@ public Properties getRuntimeProperties() {
...
 		runtimeProperties.setProperty(ModuleConstants.IGNORE_CORE_MODULES_PROPERTY, "true");
 		
-		try {
-			File tempappdir = File.createTempFile("appdir-for-unit-tests-", "");
-			tempappdir.delete(); // so we can make it into a directory
-			tempappdir.mkdir(); // turn it into a directory
-			tempappdir.deleteOnExit(); // clean up when we're done with tests
-			
-			runtimeProperties.setProperty(OpenmrsConstants.APPLICATION_DATA_DIRECTORY_RUNTIME_PROPERTY, tempappdir
+			runtimeProperties.setProperty(OpenmrsConstants.APPLICATION_DATA_DIRECTORY_RUNTIME_PROPERTY, tempappdir.getRoot()
```
[the remark this answers is not in the document]
Comment (jwnasambu) [reply api/src/test/java/org/openmrs/test/BaseContextSensitiveTest.java:360]: @teleivo thanks for reviewing my work! Kindly help me understand the tab that you are referring too. I wouldn't want to send you to laughter by telling you what I did that I thought I had sorted this "tab" issue.

[the remark this answers is not in the document]
Comment (teleivo) [reply api/src/test/java/org/openmrs/test/BaseContextSensitiveTest.java:360]: I made the comments in this file on the lines that you need to remove the tab. In general if you remove a block like if/else,try/catch,... and you keep code that was previously inside that block you will need to remove one level if nesting. So here you removed the try/catch and keep 2 lines of code. These 2 lines need to be unnested, thus one tab needs to be removed.

Comments on the PR are mostly attached to lines of code. So you can take that information into account when trying to understand the comment.

**Comment by ibacher** [comment]
@jwnasambu @teleivo So the issue here turns out to be much more subtle than I thought. It's not just that we weren't guaranteeing that the application data directory was cleaned up, but rather we seem to be caching the initial application data directory in the [HibernateSessionFactoryBean](https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/api/db/hibernate/HibernateSessionFactoryBean.java), which seems to only be called once for all of the unit tests run. This class sets a number of properties, but the most important one for this context is the `hibernate.search.default.indexBase`, which defines where Hibernate Search tells Lucene to find it's index. This property is set to `%APPLICATION_DATA_DIRECTORY%/lucene/indexes` which [gets set on the first run-through](https://github.com/openmrs/openmrs-core/blob/899095a11cc88bca9f3a01f985fe9d06962c9d7b/api/src/main/java/org/openmrs/api/db/hibernate/HibernateSessionFactoryBean.java#L167) and never changed. For some reason, this construct only tries to create the directory once, and after that it fails, hence the failed build.

I'm not really sure what the path forward is on this one, but it seems like we might be stuck with the hacky thing that kind of works and an AtExit runner...

**Comment by jwnasambu** [comment]
@ibacher  and @teleivo thanks so much for guiding me on this. Am sorry you haven't seen any changes pushed because am still running into one error after another .

**Comment by teleivo** [comment]
thank you @ibacher for your analysis! I need time to think more deeply about this. I assume that this bug is not an urgent one. It was likely around for quite some time. I therefore would like us to take more time to see if there is an alternative to the atExit runner.

**Pull request closed without merging**